### ISSUE TITLE
Android mobile web authentication not redirecting back to the app

### ISSUE THREAD

**Issue by SangCrowdstar** [pinterest/android-pdk#26]
Hi guys,
I'm having trouble authenticating Pinterest permissions on Android. It works fine with Pinterest app installed but without it, it takes me to a web browser, I hit ok but it stays there and doesn't redirect back to my app. This is my manifest:

```
<activity android:name="com.something.somethingelse.AndroidInterface.NanigansWrapper" android:label="@string/app_name">
      <intent-filter>
        <action android:name="android.intent.action.MAIN" />
        <category android:name="android.intent.category.LAUNCHER" />
	  </intent-filter>
      <intent-filter>
      	<data android:scheme="@string/app_name" android:host="open"/>
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <data android:scheme="pdk12345435464565443543" />
        <!-- data android:scheme="com.crowdstar.designhome" /-->
      </intent-filter>
      <meta-data android:name="unityplayer.UnityActivity" android:value="true" />
</activity>
```

Any idea ?

Thanks,

**Comment by lordscales91** [comment]
To properly see the code type three backticks (`) followed by xml on a new line. paste your manifest and then put another three backticks on a new line to produce an output like this:
```xml
<activity
            android:name=".MainActivity"
            android:label="@string/app_name" >
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
            <intent-filter>
                <action android:name="android.intent.action.VIEW" />
                <category android:name="android.intent.category.DEFAULT" />
                <category android:name="android.intent.category.BROWSABLE" />
                <data android:scheme="pdk[YOUR-APP-ID]" />
            </intent-filter>
</activity>
```
This is how I have my manifest.
Remember to add the following lines on the `onCreate` method of that activity
```java
PDKClient.configureInstance(this, "YOUR-APP-ID");
PDKClient.getInstance().onConnect(this); // This one completes the login extracting 
                                    // the access token from the intent data
```

**Comment by SangCrowdstar** [comment]
Hi, thanks for the tip on inserting codes. I have those 2 lines on onCreate method but it still didn't work

**Comment by SangCrowdstar** [comment]
I finally got it to redirect back to my app by removing `<data android:scheme="@string/app_name" android:host="open"/>` from intent filter. But now onActivityResult doesn't get called so I'm not getting any data back.

**Comment by lordscales91** [comment]
@SangCrowdstar I think you are a bit confused about how the authentication works.

If the Pinterest app is installed, the PDK calls it with a `startActivityForResult(intent);` then after authorizing your `onActivityResult` is called. However, if it's not installed the process is a bit different. First the user is redirected to the web browser for the authorization, then the user is redirected back to your app with the access token in the Uri. But, the `onActivityResult` will not be called at all, if you are expecting it to be called you are wrong.

If you want to get notified you currently have two options:
* check for the presence of the access token in the Uri of the calling intent (this will be a signal that your app is authorized)
```java
if (Intent.ACTION_VIEW.equals(activity.getIntent().getAction())) {
            Uri uri = activity.getIntent().getData();
            if (uri != null && uri.toString().contains("pdk" + _clientId + "://")) {
                    if(uri.getQueryParameter("access_token") != null) {
                             // The app is authorized. You can start calling API method now
                    }
            }
}
```
* Implement `onSuccess()` in the PDKCallback
```java
// In some method
PDKClient.getInstance().login(this, scopes, new PDKCallback() {
			@Override
			public void onSuccess(PDKResponse response) {
				// Log.d("PDKLogin", response.getData().toString());
				onLoginSuccess(response);
			}
// on the Activity
protected void onLoginSuccess(PDKResponse response) {
      // You can use response.getUser() here to get the logged in user. 
}
```
Alternatively, you can use my [fork](https://github.com/lordscales91/android-pdk/tree/dev_master) and use the method that I've implemented to perform a silent login. Ideally Pinterest staff should merge my PR, but it can take a long time, until then feel free to use my fork at your own risk.

**Comment by SangCrowdstar** [comment]
@lordscales91 I did implement onSuccess and onFailure in the PDKcallback but it doesn't get called when the web authentication is done and redirected back to the app.

I noticed they use startActivity() for web authentication, that's why onActivityResult never gets called. Also, the URI checking is done inside onConnect() but onConnect() is called on onCreate.

So when it gets back to the app, how do I trigger the uri checking like u suggested?. This is my PDKcallback

```
PDKClient.getInstance().login(this, scopes, new PDKCallback() {
			    @Override
			    public void onSuccess(PDKResponse response) {
			        android.util.Log.w("Pinterest", "Pinterest Ask Permission Success" + response.getData().toString());
			        //user logged in, use response.getUser() to get PDKUser object
			       //PDKUser foundUser = response.getUser();
			        String boardIdList = response.getBoardList().toString();
			        UnityPlayer.UnitySendMessage("PinterestObject","OnPinterestSuccess", response.getData().toString()+" "+boardIdList);
			        //CreatePinterestBoard(boardName, boardDesc);
			        //pinterestUsername = foundUser.getUid();
			    }

			    @Override
			    public void onFailure(PDKException exception) {
			        //android.util.Log.w("Pinterest", "Pinterest Exception:"+exception.getDetailMessage());
			        android.util.Log.w("Pinterest", "Pinterest Ask Permission Failed" + exception.getDetailMessage());
			        //android.util.Log.w("Nanigans", "Failed: "+ exception.getDetailMessage());
			        UnityPlayer.UnitySendMessage("PinterestObject", "OnPinterestFail", "DSHomeMessageError: "+exception.getDetailMessage());
			        //return notFoundString;
			    }
```

PS: I'm a Unity developer, and this is the first time I implement an Android SDK into Unity project. So my knowledge on Android development is quite limited.

**Comment by SangCrowdstar** [comment]
btw just found out, hitting "Okay" in the web authentication doesn't event grant the permissions (I checked in my profile settings)

**Comment by lordscales91** [comment]
> Also, the URI checking is done inside onConnect() but onConnect() is called on onCreate.
So when it gets back to the app, how do I trigger the uri checking like u suggested?

Yeah, I don't blame you. I think the Pinterest SDK is still very immature. The problem is that the login workflow is not correctly implemented, as I mentioned in #25. Furthermore, due to the lifecycle of an Android app the callback is not properly called. I will try to describe what happens step by step.

* First, you trigger the login by calling the method.
* Your app goes to the background and the web browser is placed on the top of the app stack.
* The browser calls your app and it's placed again on top. But it's not the same instance, the old one is still back in the stack (you will notice it after pressing the back button several time). Therefore, the callback is never called.

You will to trigger the login again or attempt to call the API.
I have an idea, try to place something like this on your `onCreate`
```java
PDKClient.configureInstance(this, "YOUR-APP-ID");
PDKClient.getInstance().onConnect(this); // This one completes the login extracting 
                                    // the access token from the intent data
PDKClient.getInstance().getMe(new PDKCallback() { // You can try it with a different method
     @Override
     public void onSuccess(PDKResponse response) { 
            storeUser(response.getUser());
      }
});
// define the method in your Activity
protected void storeUser(PDKUser user) {
    // do something with the user
}
```
Basically the idea is to just try to call the API, if it doesn't succeeds, then the app is not authorized yet. Otherwise the client is ready to use.

> btw just found out, hitting "Okay" in the web authentication doesn't event grant the permissions (I checked in my profile settings).

You should contact Pinterest staff for advice because that is really weird. I'm sorry but I can't help you more. I will release soon the source code of an app I have been working on using my forked version of the PDK. I will let you know so you can check it out.

**Comment by SangCrowdstar** [comment]
Hi, if you can fix the weblogin function in your fork so it'll return a response when it switches back to the app. That'd be a great help. Looking forward to your reply.

Thanks.

**Comment by lordscales91** [comment]
Hi @SangCrowdstar . Sadly, it's not possible to get "notified" with web login. This because of the way the back task queue and activity stack works. You should check this: https://developer.android.com/guide/components/tasks-and-back-stack.html?hl=en#ManagingTasks

The problem is that the app never "switches back", instead a new instance of the Activity is placed on the top. So any callback will just not reach anywhere. You could try to change that behaviour by defining a launch mode. But as stated in the docs this is not recommended and probably will cause some undesired side effects.

Instead I invite you to check my little [open source project](https://github.com/lordscales91/pinit-android) and use my forked library. Take a look at these lines [MainActivity.java#L72-L90](https://github.com/lordscales91/pinit-android/blob/4e82fe4247ecb5f502b384adb57e1a94d17fdfa6/src/io/github/lordscales91/pinit/MainActivity.java#L72-L90)

You can implement your own logic in the `onLoginSuccess()` method. Such as calling the API again to get the user boards.

**Comment by Zehhu** [comment]
I'm going to link this to #25, we have some plans to fix the login flow and redo it. Stay tuned.
